src/lib.rs

```diff
@@ -282,12 +282,33 @@ pub extern "C" fn setsockopt(
     optval: *const c_void,
     optlen: socklen_t,
 ) -> c_int {
-}
+}*/
 
 #[no_mangle]
-pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {}
+pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {
+    let lock = match INDEX.get(fd) {
+        Some(s) => s,
+        None => return next!(listen(fd, n)),
+    };
 
-#[no_mangle]
+    let mut sock = lock.write().unwrap();
+
+    match *sock {
```

Comment:
Personally I prefer an `if let` for the outer match:

```rust
if let Socket::Bound(p) = *sock {
    match next!(listen(fd(n)) {
        /* ... */
    }
}
error(libc::EBADFD, -1)
```

Saves you a line and more importantly a layer of indentation.

Reply:
Agreed.